bootloader/piboot.go

```diff
@@ -0,0 +1,431 @@
+)
+
+const (
+	pibootCfgFilename = "piboot.conf"
+	pibootPartFolder  = "/piboot/ubuntu/"
+	// TODO The ubuntu-seed folder should be eventually passed
+	// around when creating the bootloader
+	runMntDir     = "/run/mnt/"
+	ubuntuSeedDir = runMntDir + "ubuntu-seed/"
+)
+
+func getSeedPartDirRuntime() string {
+	return ubuntuSeedDir
+}
+
+func getRebootParamPathRuntime() string {
+	return "/run/systemd/reboot-param"
+}
+
+// Get absolute dirs via variables so we can mock in tests
+var (
+	getSeedPartDir     = getSeedPartDirRuntime
+	getRebootParamPath = getRebootParamPathRuntime
+)
+
+type piboot struct {
+	rootdir string
+	basedir string
+}
+
+func (p *piboot) setDefaults() {
+	p.basedir = "/boot/piboot/"
+}
+
+func (p *piboot) processBlOpts(blOpts *Options) {
+	if blOpts == nil {
+		return
+	}
+
+	switch {
+	case blOpts.Role == RoleRecovery || blOpts.NoSlashBoot:
+		if !blOpts.PrepareImageTime {
+			p.rootdir = getSeedPartDir()
+		}
+		// RoleRecovery or NoSlashBoot imply we use
+		// the environment file in /piboot/ubuntu as
+		// it exists on the partition directly
+		p.basedir = pibootPartFolder
+	}
+}
+
+// newPiboot creates a new Piboot bootloader object
+func newPiboot(rootdir string, blOpts *Options) Bootloader {
+	p := &piboot{
+		rootdir: rootdir,
+	}
+	p.setDefaults()
+	p.processBlOpts(blOpts)
+	return p
+}
+
+func (p *piboot) Name() string {
+	return "piboot"
+}
+
+func (p *piboot) dir() string {
+	if p.rootdir == "" {
+		panic("internal error: unset rootdir")
+	}
+	return filepath.Join(p.rootdir, p.basedir)
+}
+
+func (p *piboot) envFile() string {
+	return filepath.Join(p.dir(), pibootCfgFilename)
+}
+
+// piboot enabled if env file exists
+func (p *piboot) Present() (bool, error) {
+	return osutil.FileExists(p.envFile()), nil
+}
+
+// Variables stored in ubuntu-seed:
+//   snapd_recovery_system
+//   snapd_recovery_mode
+//   snapd_recovery_kernel
+// Variables stored in ubuntu-boot:
+//   kernel_status
+//   snap_kernel
+//   snap_try_kernel
+//   snapd_extra_cmdline_args
+//   snapd_full_cmdline_args
+//   recovery_system_status
+//   try_recovery_system
+func (p *piboot) SetBootVars(values map[string]string) error {
+	env, err := ubootenv.OpenWithFlags(p.envFile(), ubootenv.OpenBestEffort)
+	if err != nil {
+		return err
+	}
+
+	// Set when we change a boot env variable, to know if we need to save the env
+	dirtyEnv := false
+	// Flag to know if we need to write piboot's config.txt or tryboot.txt
+	reconfigBootloader := false
+	for k, v := range values {
+		// already set to the right value, nothing to do
+		if env.Get(k) == v {
+			continue
+		}
+		env.Set(k, v)
+		dirtyEnv = true
+		// Cases that change the bootloader configuration
+		if k == "snapd_recovery_mode" || k == "kernel_status" {
+			reconfigBootloader = true
+		}
+		if k == "snap_try_kernel" && v == "" {
+			// refresh (ok or not) finished, remove tryboot.txt
+			// (os_prefix in config.txt will be changed now in
+			// loadAndApplyConfig in the ok case)
```

Comment:
this should make clear that there is no ordering concern here since os_prefix is always safe from config.txt, the dangerous "trying" os_prefix is the one set in tryboot.txt, and that will never be used accidentally since tryboot.txt is only read when the special boot register is set.

When I initially read this again I wondered if there was an issue with removing tryboot.txt before we set os_prefix inside config.txt and after reading your piboot.md was reminded how it worked, but it would be nice to just make that super obvious from inside the code

Reply:
I've added some clarification on this